### PR TITLE
Add Text.substring to allow for an easy short hand of Text.take (start.up_to end)

### LINKED ISSUE #7876: Add `Text.substring` to allow for an easy short hand of `Text.take (start.up_to end)`
Review Java, C#, Python, Alteryx and Excel should it be length or end as second argument...

Add a short hand function to get a sub-string from a Text value.

### PR DESCRIPTION
**Pull request by Cassandra-Clark**
### Pull Request Description

Implemented substring for #7876. Also simplified get function.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the guide
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 1]
```diff
@@ -1791,6 +1786,35 @@ Text.parse_time_of_day self format:Date_Time_Formatter=Date_Time_Formatter.iso_t
 Text.parse_time_zone : Time_Zone ! Time_Error
 Text.parse_time_zone self = Time_Zone.parse self
 
+## ALIAS mid, slice, substring
+   GROUP Selections
+   Creates a new Text by selecting the specified range of the input.
+
+   Arguments:
+   'start' - the character position to start the substring at. 0-based index. Has support for negative values
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]:
```suggestion
   'start' - the character position to start the substring at. 0-based index. Has support for negative values.
```

File: test/Tests/src/Data/Text_Spec.enso [hunk 2]
```diff
@@ -1647,5 +1647,17 @@ spec =
...
+        Test.specify "should work with all examples" <|
+            "Hello World!".substring 3 2 . should_equal "lo"
+            "Hello World!".substring 5 . should_equal " World!"
+            "Hello World!".substring 5 7 . should_equal " World!"
+            "Hello World!".substring -7 3 . should_equal " Wo"
+
+        Test.specify "should error on length less than 0" <|
+            "Hello World!".substring 5 -2 . should_fail_with Illegal_Argument
+
+        Test.specify "should error if negative start index is too large"
```
Comment (GregoryTravis) [inline test/Tests/src/Data/Text_Spec.enso:1660]: Please add a test for positive start index is too large, and length is too long.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 3]
```diff
@@ -1791,6 +1786,35 @@ Text.parse_time_of_day self format:Date_Time_Formatter=Date_Time_Formatter.iso_t
...
+   'length' - the number of characters to include from start.
+
+   Returns:
+   Part of the selected Text starting at start and stopping after length.
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso:1798]: What happens if start+length goes past the end of the string?

**Review by GregoryTravis — CHANGES_REQUESTED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 3 above)
Comment (Cassandra-Clark) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso:1798]: as long as length is >= 0 it shouldn't matter what size it is, as any length or start+length that is longer than the end of the string will simply return the full substring between start and the last character of the string.
image

File: test/Tests/src/Data/Text_Spec.enso (hunk 2 above)
Comment (Cassandra-Clark) [reply test/Tests/src/Data/Text_Spec.enso:1660]: size of length should not matter as implemented, will add a test for positive start index too large.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 4]
```diff
@@ -1791,6 +1786,35 @@ Text.parse_time_of_day self format:Date_Time_Formatter=Date_Time_Formatter.iso_t
 Text.parse_time_zone : Time_Zone ! Time_Error
 Text.parse_time_zone self = Time_Zone.parse self
 
+## ALIAS mid, slice, substring
+   GROUP Selections
+   Creates a new Text by selecting the specified range of the input.
+
+   Arguments:
+   'start' - the character position to start the substring at. 0-based index. Has support for negative values.
+   'length' - the number of characters to include from start.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Please note to use consistent style for the `Arguments:` section, like other methods.

```suggestion
   - start: the character position to start the substring at. 0-based index. Has support for negative values.
   - length: the number of characters to include from start.
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 5]
```diff
@@ -1791,6 +1786,35 @@ Text.parse_time_of_day self format:Date_Time_Formatter=Date_Time_Formatter.iso_t
...
+         "Hello World!".substring 5 7 == " World!"
+         "Hello World!".substring -7 3 == " Wo"
+
+Text.substring : Integer -> Integer -> Range ! Index_Out_Of_Bounds
+Text.substring self start:Integer length:Integer=self.length =
+    if length < 0 then Error.throw (Illegal_Argument.Error "length must be greater than or equal to 0") else
+        if length == 0 then "" else
+            self_length = self.length
+            fixed_start = if start < 0 then start + self_length else start
+            if fixed_start < 0 then (Error.throw (Index_Out_Of_Bounds.Error start self.length)) else
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: We can use the length cached just above
```suggestion
            if fixed_start < 0 then (Error.throw (Index_Out_Of_Bounds.Error start self_length)) else
```

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 6]
```diff
@@ -1791,6 +1786,35 @@ Text.parse_time_of_day self format:Date_Time_Formatter=Date_Time_Formatter.iso_t
 Text.parse_time_zone : Time_Zone ! Time_Error
 Text.parse_time_zone self = Time_Zone.parse self
 
+## ALIAS mid, slice, substring
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso:1789]: Regarding: `slice`

We already _do have_ a `slice` method on `Vector`:
```
    ## GROUP Selections
       Creates a new vector with the skipping elements until `start` and then
       continuing until `end` index.

       Arguments:
       - start: The index of the first element to include.
       - end: The index to stop slicing at.

       > Example
         Remove the first 2 elements then continue until index 5 from the vector.

             [1, 2, 3, 4, 5, 6, 7, 8].slice 2 5 == [3, 4, 5]
    slice : Integer -> Integer -> Vector Any
    slice self start end = @Builtin_Method "Vector.slice"
```

I'm slightly worried that with this alias users can get confused as to what the second argument should be - in `Vector.slice` that is `end` but in `Text.substring` (also aliased by `slice`) it is `length`.

Shouldn't we unify this? I was thinking that `length` argument makes more sense for `substring`, but actually Java's `String::substring` also takes `endIndex` and not `length`.

**Review by radeusgd — APPROVED** [review]
Looks great!

I have one thought regarding `length` vs `end` here, wondering if we should try to uniformize this with other similar methods (but also happy to hear why we should _not_ 🙂).

Anyway, I think we can discuss this asynchronously and if necessary do a separate follow up PR - no need to block this one.

**Comment by radeusgd** [comment]
And a small nitpick: please ensure to name the PR before merging it, as the name will be part of the commit history so it's good to make it something clearly readable - e.g. it could be the same/similar to the ticket name.

> Edit: typo 😅

**Comment by jdunkerley** [comment]
Having discussed with @radeusgd - lets mark the `slice` method on Vector/Array as PRIVATE to avoid confusion.
We can remove the alias to slice on the new substring method as well.

@Cassandra-Clark please make these changes and add a changelog entry then we should be ready to mark this PR as ready to merge.

**Review by jdunkerley — APPROVED** [review]
Missing adding the link within the ChangeLog.md otherwise good.

**Pull request merged**